# Project 3

- **Dataset(s) to be used:** [[Electric Vehicle Data in Washington State](https://data.wa.gov/demographics/Washington-State-Cities-and-Counties/g2kf-7usg/about_data)]
[[Presidential Election Results in Washington State](https://results.vote.wa.gov/results/20241105/export.html)]
[[Population by County in Washington State](https://ofm.wa.gov/washington-data-research/population-demographics/population-estimates/april-1-official-population-estimates)]

- **Analysis question:** [Do Washington State counties that voted for Kamala Harris in the 2024 election have more EVs per capita?]
- **Columns that will (likely) be used:**
  - [County]
  - [Jurisdiction]
  - [Candidate]
  - [PercentageOfTotalVotes]
  - [2024 Population Estimate]
- (If you're using multiple datasets) **Columns to be used to merge/join them:**
  - [Dataset 1] [County]
  - [Dataset 2] [County]
  - [Dataset 3] [County]
- **Hypothesis**: [Percentage of voters in Washington State counties who voted for Kamala Harris in the 2024 presidential election is associated with a higher count of EVs per capita.]
- **Site URL:** [URL from Publish section]

### Import Packages:

In [147]:
import plotly.io as pio

pio.renderers.default = "vscode+jupyterlab+notebook_connected"

In [148]:
import pandas as pd
import plotly.express as px

### Read in EV Data:

In [149]:
ev_data = pd.read_csv("Electric_Vehicle_Population_Data_20241205 (2).csv")
ev_data

VIN (1-10)    County            City State  Postal Code  Model Year  \
0       1N4AZ0CP8D      King       Shoreline    WA      98177.0        2013   
1       5YJSA1E45K      King         Seattle    WA      98112.0        2019   
2       WVGUNPE28M    Kitsap          Olalla    WA      98359.0        2021   
3       JTDKARFP6H  Thurston         Olympia    WA      98501.0        2017   
4       1FADP5CU9G  Thurston         Olympia    WA      98506.0        2016   
...            ...       ...             ...   ...          ...         ...   
216767  1G1RB6E44D   Douglas  East Wenatchee    WA      98802.0        2013   
216768  KNDCS3LF9R   Whatcom      Bellingham    WA      98229.0        2024   
216769  7SAYGAEE9R      King         Redmond    WA      98052.0        2024   
216770  1G1RB6E49D    Pierce      Gig Harbor    WA      98329.0        2013   
216771  5YJ3E1EA1P    Pierce     Bonney Lake    WA      98391.0        2023   

              Make        Model                   Electric Vehicle Type  \
0           NISSAN         LEAF          Battery Electric Vehicle (BEV)   
1            TESLA      MODEL S          Battery Electric Vehicle (BEV)   
2       VOLKSWAGEN         ID.4          Battery Electric Vehicle (BEV)   
3           TOYOTA  PRIUS PRIME  Plug-in Hybrid Electric Vehicle (PHEV)   
4             FORD        C-MAX  Plug-in Hybrid Electric Vehicle (PHEV)   
...            ...          ...                                     ...   
216767   CHEVROLET         VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   
216768         KIA         NIRO  Plug-in Hybrid Electric Vehicle (PHEV)   
216769       TESLA      MODEL Y          Battery Electric Vehicle (BEV)   
216770   CHEVROLET         VOLT  Plug-in Hybrid Electric Vehicle (PHEV)   
216771       TESLA      MODEL 3          Battery Electric Vehicle (BEV)   

        Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0                 Clean Alternative Fuel Vehicle Eligible            75.0   
1                 Clean Alternative Fuel Vehicle Eligible           270.0   
2       Eligibility unknown as battery range has not b...             0.0   
3                   Not eligible due to low battery range            25.0   
4                   Not eligible due to low battery range            19.0   
...                                                   ...             ...   
216767            Clean Alternative Fuel Vehicle Eligible            38.0   
216768            Clean Alternative Fuel Vehicle Eligible            33.0   
216769  Eligibility unknown as battery range has not b...             0.0   
216770            Clean Alternative Fuel Vehicle Eligible            38.0   
216771  Eligibility unknown as battery range has not b...             0.0   

        Base MSRP  Legislative District  DOL Vehicle ID  \
0             0.0                  32.0       125450447   
1             0.0                  43.0       101662900   
2             0.0                  26.0       272118717   
3             0.0                  22.0       349372929   
4             0.0                  22.0       171625653   
...           ...                   ...             ...   
216767        0.0                  12.0       122822822   
216768        0.0                  40.0       267143887   
216769        0.0                  48.0       274988388   
216770        0.0                  26.0       117353064   
216771        0.0                  31.0       240754006   

                   Vehicle Location  \
0       POINT (-122.36498 47.72238)   
1       POINT (-122.30207 47.64085)   
2       POINT (-122.54729 47.42602)   
3       POINT (-122.89166 47.03956)   
4       POINT (-122.87741 47.05997)   
...                             ...   
216767  POINT (-120.29473 47.41515)   
216768   POINT (-122.45486 48.7449)   
216769  POINT (-122.13158 47.67858)   
216770   POINT (-122.6658 47.38336)   
216771  POINT (-122.17144 47.19175)   

                                         Electric Utili

We want the count of registered EVs per county, so we will use groupby:

In [150]:
ev_by_county = ev_data.groupby("County").size().reset_index(name = "Registered EVs")
ev_by_county

County  Registered EVs
0           Ada               2
1         Adams              72
2       Alameda               5
3     Albemarle               2
4    Alexandria               3
..          ...             ...
200  Williamson               1
201      Yakima            1340
202        Yolo               2
203        York               1
204        Yuba               1

[205 rows x 2 columns]

Oh no, it looks like this data is incorrect. It says all of the counties are in Washington state, but some of these counties are outside Washington State. We know this because Washington State only has 39 counties. We will have to filter the counties down to just Washington counties. For this, we can use the population data, so lets read it in

### Read in Population Data:

In [151]:
population_data = pd.read_csv("ofm_april1_population_final.csv")
population_data.head(20)

Line Filter  County                  Jurisdiction 2020 Population Census  \
0     1      1   Adams                  Adams County                 20,613   
1     2      2   Adams   Unincorporated Adams County                  9,472   
2     3      3   Adams     Incorporated Adams County                 11,141   
3     4      4   Adams                        Hatton                     79   
4     5      4   Adams                          Lind                    535   
5     6      4   Adams                       Othello                  8,549   
6     7      4   Adams                     Ritzville                  1,767   
7     8      4   Adams                     Washtucna                    211   
8     9      .       .                             .                      .   
9    10      1  Asotin                 Asotin County                 22,285   
10   11      2  Asotin  Unincorporated Asotin County                 13,920   
11   12      3  Asotin    Incorporated Asotin County                  8,365   
12   13      4  Asotin                        Asotin                  1,204   
13   14      4  Asotin                     Clarkston                  7,161   
14   15      .       .                             .                      .   
15   16      1  Benton                 Benton County                206,873   
16   17      2  Benton  Unincorporated Benton County                 36,556   
17   18      3  Benton    Incorporated Benton County                170,317   
18   19      4  Benton                   Benton City                  3,479   
19   20      4  Benton                     Kennewick                 83,921   

   2021 Population Estimate¹ 2022 Population Estimate  \
0                     20,900                   21,100   
1                      9,575                    9,575   
2                     11,325                   11,525   
3                         80                       80   
4                        535                      535   
5                      8,725                    8,920   
6                      1,775                    1,780   
7                        210                      210   
8                          .                        .   
9                     22,500                   22,600   
10                    14,070                   14,165   
11                     8,430                    8,435   
12                     1,215                    1,220   
13                     7,215                    7,215   
14                         .                        .   
15                   209,400                  212,300   
16                    36,760                   37,445   
17                   172,640                  174,855   
18                     3,500                    3,710   
19                    84,620                   85,320   

   2023 Population Estimate 2024 Population Estimate  
0                    21,200                   21,475  
1                     9,585                    9,628  
2                    11,615                   11,847  
3                        80                      167  
4                       535                      535  
5                     9,005                    9,145  
6                     1,785                    1,790  
7                       210                      210  
8                         .                        .  
9                    22,650                   22,725  
10                   14,205                   14,260  
11                    8,445                    8,465  
12                    1,225                    1,225  
13                    7,220                    7,240  
14                        .                        .  
15                  215,500                  217,850  
16                   37,615                   37,805  
17                  177,885                  180,045  
18                    3,810                    3,845  
19                   86,470                  

Let's filter out the rows we don't need

In [152]:
filtered_population_data = population_data[population_data["Jurisdiction"].str.contains("County", case=False, na=False)]
filtered_population_data.head(20)

Line Filter    County                    Jurisdiction  \
0     1      1     Adams                    Adams County   
1     2      2     Adams     Unincorporated Adams County   
2     3      3     Adams       Incorporated Adams County   
9    10      1    Asotin                   Asotin County   
10   11      2    Asotin    Unincorporated Asotin County   
11   12      3    Asotin      Incorporated Asotin County   
15   16      1    Benton                   Benton County   
16   17      2    Benton    Unincorporated Benton County   
17   18      3    Benton      Incorporated Benton County   
24   25      1    Chelan                   Chelan County   
25   26      2    Chelan    Unincorporated Chelan County   
26   27      3    Chelan      Incorporated Chelan County   
33   34      1   Clallam                  Clallam County   
34   35      2   Clallam   Unincorporated Clallam County   
35   36      3   Clallam     Incorporated Clallam County   
40   41      1     Clark                    Clark County   
41   42      2     Clark     Unincorporated Clark County   
42   43      3     Clark       Incorporated Clark County   
52   53      1  Columbia                 Columbia County   
53   54      2  Columbia  Unincorporated Columbia County   

   2020 Population Census 2021 Population Estimate¹ 2022 Population Estimate  \
0                  20,613                    20,900                   21,100   
1                   9,472                     9,575                    9,575   
2                  11,141                    11,325                   11,525   
9                  22,285                    22,500                   22,600   
10                 13,920                    14,070                   14,165   
11                  8,365                     8,430                    8,435   
15                206,873                   209,400                  212,300   
16                 36,556                    36,760                   37,445   
17                170,317                   172,640                  174,855   
24                 79,141                    80,000                   80,650   
25                 32,507                    33,140                   33,460   
26                 46,634                    46,860                   47,190   
33                 77,155                    77,750                   77,625   
34                 45,836                    46,170                   45,850   
35                 31,319                    31,580                   31,775   
40                503,311                   513,100                  520,900   
41                233,048                   236,200                  237,650   
42                270,263                   276,900                  283,250   
52                  3,952                     3,950                    3,950   
53                  1,385                     1,380                    1,385   

   2023 Population Estimate 2024 Population Estimate  
0                    21,200                   21,475  
1                     9,585                    9,628  
2                    11,615                   11,847  
9                    22,650                   22,725  
10                   14,205                   14,260  
11                    8,445                    8,465  
15                  215,500                  217,850  
16                   37,615                   37,805  
17                  177,885                  180,045  
24                   81,500                   82,300  
25                   33,875                   34,320  
26                   47,625                   47,980  
33                   78,075                   78,550  
34                   46,090                   46,320  
35                   31,985                   32,230  
40                  527,400                  536,300  
41                  240,155                  243,830  
42                  287,245                  292,470  
52              

In [153]:
filtered_population_data = filtered_population_data[~filtered_population_data["Jurisdiction"].str.contains("incorporated", case=False, na=False)].reset_index()
filtered_population_data.head(40)

index Line Filter        County         Jurisdiction  \
0       0    1      1         Adams         Adams County   
1       9   10      1        Asotin        Asotin County   
2      15   16      1        Benton        Benton County   
3      24   25      1        Chelan        Chelan County   
4      33   34      1       Clallam       Clallam County   
5      40   41      1         Clark         Clark County   
6      52   53      1      Columbia      Columbia County   
7      58   59      1       Cowlitz       Cowlitz County   
8      67   68      1       Douglas       Douglas County   
9      77   78      1         Ferry         Ferry County   
10     82   83      1      Franklin      Franklin County   
11     90   91      1      Garfield      Garfield County   
12     95   96      1         Grant         Grant County   
13    114  115      1  Grays Harbor  Grays Harbor County   
14    127  128      1        Island        Island County   
15    134  135      1     Jefferson     Jefferson County   
16    139  140      1          King          King County   
17    182  183      1        Kitsap        Kitsap County   
18    190  191      1      Kittitas      Kittitas County   
19    199  200      1     Klickitat     Klickitat County   
20    206  207      1         Lewis         Lewis County   
21    219  220      1       Lincoln       Lincoln County   
22    231  232      1         Mason         Mason County   
23    236  237      1      Okanogan      Okanogan County   
24    253  254      1       Pacific       Pacific County   
25    261  262      1  Pend Oreille  Pend Oreille County   
26    270  271      1        Pierce        Pierce County   
27    298  299      1      San Juan      San Juan County   
28    303  304      1        Skagit        Skagit County   
29    315  316      1      Skamania      Skamania County   
30    321  322      1     Snohomish     Snohomish County   
31    345  346      1       Spokane       Spokane County   
32    362  363      1       Stevens       Stevens County   
33    372  373      1      Thurston      Thurston County   
34    383  384      1     Wahkiakum     Wahkiakum County   
35    388  389      1   Walla Walla   Walla Walla County   
36    396  397      1       Whatcom       Whatcom County   
37    407  408      1       Whitman       Whitman County   
38    427  428      1        Yakima        Yakima County   

   2020 Population Census 2021 Population Estimate¹ 2022 Population Estimate  \
0                  20,613                    20,900                   21,100   
1                  22,285                    22,500                   22,600   
2                 206,873                   209,400                  212,300   
3                  79,141                    80,000                   80,650   
4                  77,155                    77,750                   77,625   
5                 503,311                   513,100                  520,900   
6                   3,952                     3,950                    3,950   
7                 110,730                   111,500                  112,350   
8                  42,938                    43,550                   44,000   
9                   7,178                     7,250                    7,300   
10                 96,749                    98,350                   99,750   
11                  2,286                     2,300                    2,300   
12                 99,123                   100,800                  101,800   
13                 75,636                    76,050                   76,400   
14                 86,857                    87,100                   87,700   
15                 32,977                    33,100                   33,350   
16              2,269,675                 2,287,050                2,317,700   
17                275,611                   277,700                  280,900   
18                 46,468                    45,225                   47,200   
19 

Now we have all the 39 counties above, so lets filter out the counties that are not in Washington state from the EV data file

In [154]:
filtered_ev_by_county = ev_by_county[ev_by_county["County"].isin(filtered_population_data["County"])].reset_index()
filtered_ev_by_county

index        County  Registered EVs
0       1         Adams              72
1      11        Asotin              90
2      16        Benton            2726
3      27        Chelan            1315
4      33       Clallam            1297
5      34         Clark           12938
6      36      Columbia              18
7      39       Cowlitz            1128
8      48       Douglas             463
9      57         Ferry              32
10     58      Franklin             773
11     63      Garfield               4
12     65         Grant             778
13     66  Grays Harbor             795
14     80        Island            2302
15     83     Jefferson            1162
16     87          King          110164
17     89        Kitsap            7215
18     90      Kittitas             804
19     91     Klickitat             362
20     99         Lewis             983
21    100       Lincoln              64
22    108         Mason            1026
23    129      Okanogan             331
24    133       Pacific             264
25    135  Pend Oreille              75
26    139        Pierce           17397
27    161      San Juan            1047
28    169        Skagit            2447
29    170      Skamania             224
30    171     Snohomish           26200
31    174       Spokane            5999
32    181       Stevens             263
33    185      Thurston            7938
34    191     Wahkiakum              75
35    193   Walla Walla             552
36    197       Whatcom            5244
37    198       Whitman             426
38    201        Yakima            1340

Let's graph this in a histogram:

In [155]:
fig=px.histogram(
    filtered_ev_by_county,
    x="County",
    y="Registered EVs",
    title = "Registered EVs by County in Washington State"
)
fig.show()

It's definitly a bit strange that almost all ofthe EVs are in King county (Seattle) but almost none in other counties. I was expecting King county to have the most but not by this much. Let's take another look at this histogram but per capita. We will have to merge population and EVs by county and then create a per capita column:

In [156]:
ev_and_pop_data = pd.merge(filtered_ev_by_county, filtered_population_data, on="County")
ev_and_pop_data.head()

index_x   County  Registered EVs  index_y Line Filter    Jurisdiction  \
0        1    Adams              72        0    1      1    Adams County   
1       11   Asotin              90        9   10      1   Asotin County   
2       16   Benton            2726       15   16      1   Benton County   
3       27   Chelan            1315       24   25      1   Chelan County   
4       33  Clallam            1297       33   34      1  Clallam County   

  2020 Population Census 2021 Population Estimate¹ 2022 Population Estimate  \
0                 20,613                    20,900                   21,100   
1                 22,285                    22,500                   22,600   
2                206,873                   209,400                  212,300   
3                 79,141                    80,000                   80,650   
4                 77,155                    77,750                   77,625   

  2023 Population Estimate 2024 Population Estimate  
0                   21,200                   21,475  
1                   22,650                   22,725  
2                  215,500                  217,850  
3                   81,500                   82,300  
4                   78,075                   78,550

In [157]:
ev_and_pop_data["Registered EVs per capita 2024"] = ev_and_pop_data["Registered EVs"] / ev_and_pop_data["2024 Population Estimate"]
ev_and_pop_data()

TypeError: unsupported operand type(s) for /: 'int' and 'str'

Looks like something is not an int:

In [ ]:
ev_and_pop_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 39 entries, 0 to 38
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   index_x                    39 non-null     int64 
 1   County                     39 non-null     object
 2   Registered EVs             39 non-null     int64 
 3   index_y                    39 non-null     int64 
 4   Line                       39 non-null     object
 5   Filter                     39 non-null     object
 6   Jurisdiction               39 non-null     object
 7   2020 Population Census     39 non-null     object
 8   2021 Population Estimate¹  39 non-null     object
 9   2022 Population Estimate   39 non-null     object
 10  2023 Population Estimate   39 non-null     object
 11  2024 Population Estimate   39 non-null     object
dtypes: int64(3), object(9)
memory usage: 3.8+ KB


In [ ]:
ev_and_pop_data["2024 Population Estimate"].astype(int)

ValueError: invalid literal for int() with base 10: '21,475'

In [ ]:
ev_and_pop_data["2024 Population Estimate"].unique()

array(['21,475', '22,725', '217,850', '82,300', '78,550', '536,300',
       '3,975', '113,900', '45,150', '7,350', '102,300', '2,325',
       '105,300', '77,400', '88,650', '33,700', '2,378,100', '286,100',
       '48,600', '23,450', '84,950', '11,300', '67,475', '43,200',
       '23,950', '13,850', '952,600', '18,475', '133,300', '12,075',
       '867,100', '559,400', '47,650', '307,000', '4,550', '63,375',
       '238,000', '48,750', '263,200'], dtype=object)

We need to take out the commas from population

In [ ]:
ev_and_pop_data["2024 Population Estimate"] = ev_and_pop_data["2024 Population Estimate"].str.replace(",", "")
ev_and_pop_data["2024 Population Estimate"].unique()

array(['21475', '22725', '217850', '82300', '78550', '536300', '3975',
       '113900', '45150', '7350', '102300', '2325', '105300', '77400',
       '88650', '33700', '2378100', '286100', '48600', '23450', '84950',
       '11300', '67475', '43200', '23950', '13850', '952600', '18475',
       '133300', '12075', '867100', '559400', '47650', '307000', '4550',
       '63375', '238000', '48750', '263200'], dtype=object)

In [ ]:
ev_and_pop_data["2024 Population Estimate"] = ev_and_pop_data["2024 Population Estimate"].astype(int)

Now we can make the new column:

In [ ]:
ev_and_pop_data["Registered EVs per capita 2024"] = ev_and_pop_data["Registered EVs"] / ev_and_pop_data["2024 Population Estimate"]
ev_and_pop_data.head()

index_x   County  Registered EVs  index_y Line Filter    Jurisdiction  \
0        1    Adams              72        0    1      1    Adams County   
1       11   Asotin              90        9   10      1   Asotin County   
2       16   Benton            2726       15   16      1   Benton County   
3       27   Chelan            1315       24   25      1   Chelan County   
4       33  Clallam            1297       33   34      1  Clallam County   

  2020 Population Census 2021 Population Estimate¹ 2022 Population Estimate  \
0                 20,613                    20,900                   21,100   
1                 22,285                    22,500                   22,600   
2                206,873                   209,400                  212,300   
3                 79,141                    80,000                   80,650   
4                 77,155                    77,750                   77,625   

  2023 Population Estimate  2024 Population Estimate  \
0                   21,200                     21475   
1                   22,650                     22725   
2                  215,500                    217850   
3                   81,500                     82300   
4                   78,075                     78550   

   Registered EVs per capita 2024  
0                        0.003353  
1                        0.003960  
2                        0.012513  
3                        0.015978  
4                        0.016512

Now let's graph the histogram:

In [ ]:
fig=px.histogram(
    ev_and_pop_data,
    x="County",
    y="Registered EVs per capita 2024",
    title = "Registered EVs by County in Washington State per Capita"
)
fig.show()

This makes more sense, it looks like King County just has a much larger population

Now let's take a look at the election data before we merge it with population and Registered EVs

### Read in the Election Data

In [ ]:
election_data = pd.read_csv("20241105_allcounties.csv")
election_data.head(20)

County                                          Race  \
0   Adams  Washington State Initiative Measure No. 2066   
1   Adams  Washington State Initiative Measure No. 2066   
2   Adams  Washington State Initiative Measure No. 2109   
3   Adams  Washington State Initiative Measure No. 2109   
4   Adams  Washington State Initiative Measure No. 2117   
5   Adams  Washington State Initiative Measure No. 2117   
6   Adams  Washington State Initiative Measure No. 2124   
7   Adams  Washington State Initiative Measure No. 2124   
8   Adams        United States President/Vice President   
9   Adams        United States President/Vice President   
10  Adams        United States President/Vice President   
11  Adams        United States President/Vice President   
12  Adams        United States President/Vice President   
13  Adams        United States President/Vice President   
14  Adams        United States President/Vice President   
15  Adams        United States President/Vice President   
16  Adams        United States President/Vice President   
17  Adams        United States President/Vice President   
18  Adams        United States President/Vice President   
19  Adams                    United States U.S. Senator   

                                  Candidate  \
0                                       Yes   
1                                        No   
2                                       Yes   
3                                        No   
4                                       Yes   
5                                        No   
6                                       Yes   
7                                        No   
8               Kamala D. Harris / Tim Walz   
9                Donald J. Trump / JD Vance   
10  Robert F. Kennedy Jr. / Nicole Shanahan   
11                Jill Stein / Rudolph Ware   
12       Claudia De la Cruz / Karina Garcia   
13           Rachele Fruit / Dennis Richter   
14             Joseph Kishore / Jerry White   
15             Chase Oliver / Mike ter Maat   
16            Cornel West / Melina Abdullah   
17          Shiva Ayyadurai / Crystal Ellis   
18                                 WRITE-IN   
19                           Maria Cantwell   

                                      Party  Votes  PercentageOfTotalVotes  \
0                                             3402                 65.8282   
1                                             1766                 34.1718   
2                                             2571                 49.3664   
3                                             2637                 50.6336   
4                                             2969                 57.3609   
5                                             2207                 42.6391   
6                                             2869                 55.3220   
7                                             2317                 44.6780   
8                 Democratic Party Nominees   1455                 27.1202   
9                 Republican Party Nominees   3767                 70.2144   
10             We The People Party Nominees     75                  1.3979   
11                     Green Party Nominees     11                  0.2050   
12  Socialism and Liberation Party Nominees      8                  0.1491   
13         Socialist Workers Party Nominees      1                  0.0186   
14        Socialist Equality Party Nominees      0                  0.0000   
15               Libertarian Party Nominees     20                  0.3728   
16           Justice For All Party Nominees      4                  0.0746   
17                   Independent Candidates      4                  0.0746   
18                                              20                  0.3728   
19               (Prefers Democratic Party)   1627                 31.4396   

   JurisdictionName  
0   State Executive  
1   State Executive  
2   State Executive  
3   State Executive  
4   State Executive  
5   

We only need percentage of votes for Kamala Harris for each county, so let's filter down our dataset:

In [ ]:
filtered_election_data = election_data[election_data["Candidate"].str.contains("Kamala", case=False, na=False)].reset_index()
filtered_election_data.head(40)

index        County                                    Race  \
0       8         Adams  United States President/Vice President   
1      94        Asotin  United States President/Vice President   
2     175        Benton  United States President/Vice President   
3     277        Chelan  United States President/Vice President   
4     376       Clallam  United States President/Vice President   
5     510         Clark  United States President/Vice President   
6     625      Columbia  United States President/Vice President   
7     704       Cowlitz  United States President/Vice President   
8     795       Douglas  United States President/Vice President   
9     887         Ferry  United States President/Vice President   
10    976      Franklin  United States President/Vice President   
11   1075      Garfield  United States President/Vice President   
12   1148         Grant  United States President/Vice President   
13   1235  Grays Harbor  United States President/Vice President   
14   1333        Island  United States President/Vice President   
15   1424     Jefferson  United States President/Vice President   
16   1503          King  United States President/Vice President   
17   1705        Kitsap  United States President/Vice President   
18   1799      Kittitas  United States President/Vice President   
19   1876     Klickitat  United States President/Vice President   
20   1966         Lewis  United States President/Vice President   
21   2064       Lincoln  United States President/Vice President   
22   2144         Mason  United States President/Vice President   
23   2226      Okanogan  United States President/Vice President   
24   2311       Pacific  United States President/Vice President   
25   2393  Pend Oreille  United States President/Vice President   
26   2484        Pierce  United States President/Vice President   
27   2673      San Juan  United States President/Vice President   
28   2753        Skagit  United States President/Vice President   
29   2853      Skamania  United States President/Vice President   
30   2941     Snohomish  United States President/Vice President   
31   3091       Spokane  United States President/Vice President   
32   3224       Stevens  United States President/Vice President   
33   3310      Thurston  United States President/Vice President   
34   3429     Wahkiakum  United States President/Vice President   
35   3508   Walla Walla  United States President/Vice President   
36   3593       Whatcom  United States President/Vice President   
37   3732       Whitman  United States President/Vice President   
38   3864        Yakima  United States President/Vice President   

                      Candidate                      Party   Votes  \
0   Kamala D. Harris / Tim Walz  Democratic Party Nominees    1455   
1   Kamala D. Harris / Tim Walz  Democratic Party Nominees    4082   
2   Kamala D. Harris / Tim Walz  Democratic Party Nominees   37662   
3   Kamala D. Harris / Tim Walz  Democratic Party Nominees   18397   
4   Kamala D. Harris / Tim Walz  Democratic Party Nominees   25440   
5   Kamala D. Harris / Tim Walz  Democratic Party Nominees  143206   
6   Kamala D. Harris / Tim Walz  Democratic Party Nominees     661   
7   Kamala D. Harris / Tim Walz  Democratic Party Nominees   22825   
8   Kamala D. Harris / Tim Walz  Democratic Party Nominees    7410   
9   Kamala D. Harris / Tim Walz  Democratic Party Nominees    1315   
10  Kamala D. Harris / Tim Walz  Democratic Party Nominees   11884   
11  Kamala D. Harris / Tim Walz  Democratic Party Nominees     330   
12  Kamala D. Harris / Tim Walz  Democratic Party Nominees   10806   
13  Kamala D. Harris / Tim Walz  Democratic Party Nominees   17161   
14  Kamala D. Harris / Tim Walz  Democratic Party Nominees   29595   
15  Kamala D. Harris / Tim Walz  Democratic Party Nominees   17459   
16  Kamala D. Harris / Tim Walz  Democratic Party Nominees  832606   
17  Kamala D. Harris / Tim Walz  Democratic Party Nominees   91

Now we can merge this with our population and ev data:

In [ ]:
merged_data = pd.merge(ev_and_pop_data, filtered_election_data, on="County")
merged_data.head()

index_x   County  Registered EVs  index_y Line Filter    Jurisdiction  \
0        1    Adams              72        0    1      1    Adams County   
1       11   Asotin              90        9   10      1   Asotin County   
2       16   Benton            2726       15   16      1   Benton County   
3       27   Chelan            1315       24   25      1   Chelan County   
4       33  Clallam            1297       33   34      1  Clallam County   

  2020 Population Census 2021 Population Estimate¹ 2022 Population Estimate  \
0                 20,613                    20,900                   21,100   
1                 22,285                    22,500                   22,600   
2                206,873                   209,400                  212,300   
3                 79,141                    80,000                   80,650   
4                 77,155                    77,750                   77,625   

  2023 Population Estimate  2024 Population Estimate  \
0                   21,200                     21475   
1                   22,650                     22725   
2                  215,500                    217850   
3                   81,500                     82300   
4                   78,075                     78550   

   Registered EVs per capita 2024  index  \
0                        0.003353      8   
1                        0.003960     94   
2                        0.012513    175   
3                        0.015978    277   
4                        0.016512    376   

                                     Race                    Candidate  \
0  United States President/Vice President  Kamala D. Harris / Tim Walz   
1  United States President/Vice President  Kamala D. Harris / Tim Walz   
2  United States President/Vice President  Kamala D. Harris / Tim Walz   
3  United States President/Vice President  Kamala D. Harris / Tim Walz   
4  United States President/Vice President  Kamala D. Harris / Tim Walz   

                       Party  Votes  PercentageOfTotalVotes JurisdictionName  
0  Democratic Party Nominees   1455                 27.1202          Federal  
1  Democratic Party Nominees   4082                 35.5947          Federal  
2  Democratic Party Nominees  37662                 37.4028          Federal  
3  Democratic Party Nominees  18397                 43.6041          Federal  
4  Democratic Party Nominees  25440                 52.2779          Federal

Now let's graph a scatter plot to see if our hypothesis is true, that counties with higher votes for Kamala Harris have higher EV ridership per capita

In [159]:
fig = px.scatter(
    merged_data,
    x = "Registered EVs per capita 2024",
    y = "PercentageOfTotalVotes",
    title = "Registered EVs per Capita 2024 vs Percentage of Total Votes for Kamala Harris by County",
    hover_data={'County': True},
    trendline = "ols"
)

fig.update_layout(
    yaxis_title="Percentage of Votes for Kamala Harris"
)
fig.show()

Looks like a strong positive correlation, let's see what the correlation is exactly:

In [ ]:
correlation = merged_data['Registered EVs per capita 2024'].corr(merged_data['PercentageOfTotalVotes'])
correlation

0.8792233075325347

There is a 0.88 correlation between voting for Kamala Harris and EV ridership in Washington State. 0.88 is very strong, so my hypothesis was correct!